### PR TITLE
GEOMESA-2902 Arrow dictionary encoding for list-type attributes

### PR DESCRIPTION
**Pull request by elahrvivaz**
* List items are dictionary encoded, not the list itself
* Add sorting for list-type attributes
* Move ObjectType to geomesa-utils

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/filter/ArrowFilterOptimizer.scala [hunk 1]
```diff
@@ -191,4 +196,12 @@ object ArrowFilterOptimizer extends LazyLogging {
       arrow.getReader(i).asInstanceOf[ArrowDictionaryReader].getEncoded(arrow.getIndex) == value
     }
   }
+
+  case class ArrowListDictionaryEquals(i: Int, value: Int) extends Filter {
```
Comment (jnh5y) [inline geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/filter/ArrowFilterOptimizer.scala:200]: (*Non* blocking question)

Is there a way to check for actual list equality?

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala [hunk 2]
```diff
@@ -188,16 +175,12 @@ class DeltaWriter(
 
     // set attributes in the vector
```
Comment (jnh5y) [inline geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala:176]: Does this comment need to change?  (Proof that I read things somewhat...;))

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala [hunk 3]
```diff
@@ -213,22 +196,19 @@ class DeltaWriter(
...
   import scala.collection.JavaConverters._
 
   // empty provider
-  private val provider = new MapDictionaryProvider()
-
-  private val dictionaryOrdering: Ordering[AnyRef] = new Ordering[AnyRef] {
-    override def compare(x: AnyRef, y: AnyRef): Int =
-      SimpleFeatureOrdering.nullCompare(x.asInstanceOf[Comparable[Any]], y)
-  }
+  private val Provider = new MapDictionaryProvider()
```
Comment (jnh5y) [inline geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala:211]: Any particular reason for the capital P? (No need to change it.)

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala [hunk 4]
```diff
@@ -289,16 +269,38 @@ object DeltaWriter extends StrictLogging {
           val name = fromVector.getField.getName
           val toVector = result.underlying.getChild(name)
           val transfer: (Int, Int, java.util.Map[Integer, Integer]) => Unit =
```
Comment (jnh5y) [inline geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala:271]: Just to make sure I'm understanding....

Dictionaries are paired up IntVectors.  So List[String] turns into a Dictionary + a ListVector[Int].  Further (assertion) since this is about transfers between dictionary-encoded columns, the only cases to worry with are IntVectors and List[Int]Vectors.

Is that right?

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala [hunk 5]
```diff
@@ -572,21 +599,30 @@ object DeltaWriter extends StrictLogging {
...
       val descriptor = sft.getDescriptor(name)
-      val bindings = ObjectType.selectType(descriptor)
-      val metadata = Map(SimpleFeatureVector.DescriptorKey -> SimpleFeatureTypes.encodeDescriptor(sft, descriptor))
+      val bindings = {
+        val base = ObjectType.selectType(descriptor)
+        if (descriptor.isList) { base.tail } else { base }
+      }
+      val ordering = AttributeOrdering(bindings)
+      val desc = if (descriptor.isList) { s"$name:${descriptor.getListType().getSimpleName}" } else {
+        SimpleFeatureTypes.encodeDescriptor(sft, descriptor)
```
Comment (jnh5y) [inline geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala:608]: Should this logic just be rolled into the SFTs.encodeDescriptor?  Or is that function too general for this specific, Arrow-related use?

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala [hunk 6]
```diff
@@ -832,14 +974,15 @@ object DeltaWriter extends StrictLogging {
   private class AttributeBatchMerger(
       vector: SimpleFeatureVector,
       transfers: Seq[(Int, Int) => Unit],
-      sort: ArrowAttributeReader
+      sort: ArrowAttributeReader,
+      ordering: Ordering[AnyRef]
     ) extends BatchMerger[AttributeBatchMerger](vector, transfers) {
 
-    private var value: Comparable[Any] = sort.apply(0).asInstanceOf[Comparable[Any]]
+    private var value: AnyRef = sort.apply(0)
```
Comment (jnh5y) [inline geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala:981]: The 'Comparable' drops out 'cause ....  List isn't comparable?

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DictionaryBuildingWriter.scala [hunk 7]
```diff
@@ -106,7 +108,12 @@ class DictionaryBuildingWriter(
 
     val dictionaries = attributeWriters.collect { case w: ArrowDictionaryWriter =>
       val name = s"dict-${w.dictionary.encoding.getId}"
-      val descriptor = SimpleFeatureTypes.renameDescriptor(sft.getDescriptor(w.vector.getField.getName), name)
+      val original = sft.getDescriptor(w.vector.getField.getName)
+      val descriptor = if (original.isList) {
```
Comment (jnh5y) [inline geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DictionaryBuildingWriter.scala:112]: Does this logic happen enough that it'd make sense to extract an ArrowSimpleFeatureTypes helper thingy?

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/filter/ArrowFilterOptimizer.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/filter/ArrowFilterOptimizer.scala:200]: I think geotools supports that using the filter 'match action' of 'all', but I don't think that's supported through ECQL so effectively there isn't. Maybe there is a filter function, but I'm not aware of one. We could always add one, but again I don't think there is any way to specify a list in ECQL, other than through the default converters (e.g. we could convert the string `"a,b,c"` to `List("a", "b", "c")`)

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala [hunk 8]
```diff
@@ -89,14 +93,21 @@ object SimpleFeatureArrowFileReader {
...
+      precision: SimpleFeatureEncoding): Map[String, ArrowDictionary] = {
     fields.flatMap { field =>
-      Option(field.getDictionary).toSeq.map { encoding =>
+      // check top-level dictionaries plus nested (i.e. for list-type attributes)
+      val encodings = Seq(field.getDictionary) ++ field.getChildren.asScala.map(_.getDictionary)
+      encodings.collect { case encoding if encoding != null =>
         val descriptor = SimpleFeatureTypes.createDescriptor(field.getMetadata.get(DescriptorKey))
+        val bindings = {
+          val main = ObjectType.selectType(descriptor)
+          if (descriptor.isList) { main.tail } else { main }
```
Comment (jnh5y) [inline geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala:107]: What's the type of main here?  What is its head in the case of isList==true?

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala:176]: ah, no i think it's still valid. this foreach block is setting the attributes in the vector, then they get written out to bytes after that.

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/vector/ArrowAttributeReader.scala [hunk 9]
```diff
@@ -150,22 +141,65 @@ object ArrowAttributeReader {
...
-        val dictionaryType = TypeBindings(bindings, encoding)
         vector match {
-          case v: TinyIntVector  => new ArrowDictionaryByteReader(v, dict, dictionaryType)
-          case v: SmallIntVector => new ArrowDictionaryShortReader(v, dict, dictionaryType)
-          case v: IntVector      => new ArrowDictionaryIntReader(v, dict, dictionaryType)
+          case v: TinyIntVector  => new ArrowDictionaryByteReader(v, dict)
+          case v: SmallIntVector => new ArrowDictionaryShortReader(v, dict)
+          case v: IntVector      => new ArrowDictionaryIntReader(v, dict)
+
+          case v: ListVector =>
```
Comment (jnh5y) [inline geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/vector/ArrowAttributeReader.scala:149]: Going back to my above question.  Is this covering the cases where one could have List[TinyInt/SmallInt] which are separate from Dictionary encoded situations?

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala:211]: just scala convention for singleton object vals...

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala (hunk 4 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala:271]: correct!

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala (hunk 6 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala:981]: yep! that was the error thrown when you tried to dictionary encode a list-type attribute.

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DictionaryBuildingWriter.scala (hunk 7 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DictionaryBuildingWriter.scala:112]: yeah, maybe... it's fairly simple and as I said in the other comment most places were slightly different. I could go either way, what do you think?

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala (hunk 8 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala:107]: for list types the result will be `Seq(LIST, STRING)` (or whatever the list item encoding is). so taking the tail is an easy way to get the item encoding as a seq.

Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala:107]: for e.g. a string attribute main would be `Seq(STRING)`

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/vector/ArrowAttributeReader.scala (hunk 9 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/vector/ArrowAttributeReader.scala:149]: right - if you know the number of dictionary entries up front, you can create a smaller encoding space. for the delta writer, since we don't know the sizes up front, we just encode everything as 32-bit ints. so the logic in that class doesn't have to deal with tiny/small ints.

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala [hunk 10]
```diff
@@ -88,6 +88,8 @@ class GeoJsonSerializer(sft: SimpleFeatureType) extends LazyLogging {
 
 object GeoJsonSerializer extends LazyLogging {
 
+  import org.locationtech.geomesa.utils.geotools.ObjectType
```
Comment (jnh5y) [inline geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala:91]: Why is this import needed?

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala [hunk 11]
```diff
@@ -168,11 +168,8 @@ object RichAttributeDescriptors extends Conversions {
 
     private def tryClass(value: AnyRef): Class[_] = Try(Class.forName(value.asInstanceOf[String])).getOrElse(null)
 
-    def isList: Boolean = ad.getUserData.containsKey(UserDataListType)
```
Comment (jnh5y) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala:171]: Is this change going to mess with any AttributeDescriptor data that's serialized somewhere?

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureOrdering.scala [hunk 12]
```diff
@@ -10,64 +10,66 @@ package org.locationtech.geomesa.utils.geotools
...
 import org.locationtech.geomesa.utils.collection.TieredOrdering
 import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
-import org.opengis.filter.expression.PropertyName
 import org.opengis.filter.sort.{SortBy, SortOrder}
 
 import scala.math.Ordering
 
 /**
-  * Ordering for simple features. Assumes that any attributes implement `Comparable`
+  * Ordering for simple features
```
Comment (jnh5y) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureOrdering.scala:18]: This change probably lines up perfectly with the change above between Comparable[Any] and AnyRef, right?

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala (hunk 10 above)
Comment (elahrvivaz) [reply geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala:91]: i moved ObjectType out of the geomesa-features-kryo and into geomesa-utils so that i could use it in arrow

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala (hunk 11 above)
Comment (elahrvivaz) [reply geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala:171]: it should be functionally equivalent... it should be impossible to have a list-type attribute without the user-data-list-type key, and vice-versa. it just seemed more logical to check based on the attribute type.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureOrdering.scala (hunk 12 above)
Comment (elahrvivaz) [reply geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureOrdering.scala:18]: yeah, i switched to using an ordering in DeltaWriter so that we could compare non-comparable lists, and implemented the ordering here.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/AttributeOrdering.scala [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+      Integer.compare(x.length, y.length)
+    }
+  }
+
+  private object ListOrdering {
+    def apply[T](delegate: Ordering[T]): Ordering[java.util.List[T]] = new NullOrdering(new ListOrdering(delegate))
+  }
+
+  private class ListOrdering[T](delegate: Ordering[T]) extends Ordering[java.util.List[T]] {
+    override def compare(x: java.util.List[T], y: java.util.List[T]): Int = {
```
Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/AttributeOrdering.scala:122]: @jnh5y what do you think about this as an ordering for lists? compare elements up to `min(x.length, y.length)`, then take the smaller list. there doesn't seem to be any canonical list ordering in java that i could find.

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/filter/ArrowFilterOptimizer.scala (hunk 1 above)
Comment (jnh5y) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/filter/ArrowFilterOptimizer.scala:200]: Yeah, I tried to look for use of the Match.ALL.  I believe you are correct!

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DictionaryBuildingWriter.scala (hunk 7 above)
Comment (jnh5y) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DictionaryBuildingWriter.scala:112]: Hmm....  I'm fine either way.  It isn't the end of the world if things are a little WET.  Just wanted to ask about it.

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala (hunk 8 above)
Comment (jnh5y) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala:107]: Wow.  That's sufficiently clever that I'd love to see some comments added to explain that for future us.  Is there a quick place you could add a line or two about that?

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala (hunk 10 above)
Comment (jnh5y) [reply geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala:91]: Sorry, I should I mentioned that I was asking since the same import is at the top of the file.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala (hunk 11 above)
Comment (jnh5y) [reply geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala:171]: Cool!  Just making sure we weren't changing the 'API' in some sense.

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala (hunk 8 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala:107]: it's described here: https://github.com/locationtech/geomesa/blob/main/geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/ObjectType.scala#L40-L49

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala (hunk 10 above)
Comment (elahrvivaz) [reply geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala:91]: ah, I could probably remove the top-level one now. the old ObjectTypes (which is still there but deprecated) is in this same package, and it seemed to override the top-level import so i had to also add it here to get it to compile. let me try removing the other one.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/AttributeOrdering.scala (hunk 13 above)
Comment (jnh5y) [reply geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/AttributeOrdering.scala:122]: Let's try this by an analogy...  If a string is a List[Char], it sounds like your list ordering would function the same as the ordering for String?  (More or less?)

Seems reasonable to me!

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala (hunk 10 above)
Comment (elahrvivaz) [reply geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala:90]: oh, the top-level on is the type ObjectType.ObjectType, while this one is the object ObjectType itself. they do both need to be there for it to compile. i moved the top-level one down here for consistency.

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala (hunk 8 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileReader.scala:108]: added a comment

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/AttributeOrdering.scala (hunk 13 above)
Comment (elahrvivaz) [reply geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/AttributeOrdering.scala:122]: good point!

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala (hunk 10 above)
Comment (jnh5y) [reply geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/serialization/GeoJsonSerializer.scala:90]: Ah!  *blinks*  I cannot read.  Thank you!
